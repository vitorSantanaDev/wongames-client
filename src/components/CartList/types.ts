import { GameItemProps } from 'components/GameItem/types'

export type CartListProps = {
  total: string
  hasButton?: boolean
  items: GameItemProps[]
}
