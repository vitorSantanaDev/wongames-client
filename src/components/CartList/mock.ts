import { GameItemProps } from 'components/GameItem/types'

export default [
  {
    id: '1',
    img: 'https://wallpaperaccess.com/full/3149843.jpg',
    title: 'Hoziron II',
    price: 'R$ 350,00'
  },
  {
    id: '2',
    img: 'https://wallpapers-hub.art/wallpaper-images/263763.jpg',
    title: 'Halo 5',
    price: 'R$ 255,00'
  }
] as GameItemProps[]
