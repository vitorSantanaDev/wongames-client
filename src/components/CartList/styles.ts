import { tint } from 'polished'
import styled, { css } from 'styled-components'
import media from 'styled-media-query'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-self: start;
    flex-direction: column;
    background-color: ${theme.colors.white};
  `}
`

export const CartListFooter = styled.div`
  ${({ theme }) => css`
    padding: 2rem;
    display: flex;
    justify-content: space-between;

    color: ${theme.colors.black};
    font-weight: ${theme.font.bold};
    font-size: ${theme.font.sizes.small};
    background-color: ${tint(0.2, theme.colors.lightGray)};

    ${media.greaterThan('medium')`
      padding: ${theme.spacings.small};
      font-size: ${theme.font.sizes.medium};
    `}
  `}
`

export const Total = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.primary};
  `}
`
