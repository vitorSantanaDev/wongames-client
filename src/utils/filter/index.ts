import { ItemProps } from 'components/ExploreSidebar/types'
import { ParsedUrlQueryInput } from 'querystring'

export type ParceArgs = {
  queryString: ParsedUrlQueryInput
  filterItems: Pick<ItemProps, 'type' | 'name'>[]
}

export const parseQueryStringToWhere = ({
  queryString,
  filterItems
}: ParceArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const object: any = {}

  Object.keys(queryString)
    .filter((item) => item !== 'sort')
    .forEach((key) => {
      const item = filterItems.find((item) => item.name === key)
      const isCheckbox = item?.type === 'checkbox'

      object[key] = !isCheckbox
        ? queryString[key]
        : { name_contains: queryString[key] }
    })

  return object
}

export const parseQueryStringToFilter = ({
  queryString,
  filterItems
}: ParceArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const object: any = {}

  Object.keys(queryString).forEach((key) => {
    const item = filterItems.find((item) => item.name === key)
    const isCheckbox = item?.type === 'checkbox'
    const isArray = Array.isArray(queryString[key])

    object[key] = !isArray && isCheckbox ? [queryString[key]] : queryString[key]
  })

  return object
}
