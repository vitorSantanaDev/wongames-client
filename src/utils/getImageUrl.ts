export const getImageUrl = (url: string | undefined): string | null => {
  if (process.env.NEXT_PUBLIC_IMAGE_HOST) {
    return `${process.env.NEXT_PUBLIC_IMAGE_HOST}${url}`
  }

  if (url) {
    return url
  }

  return null
}
