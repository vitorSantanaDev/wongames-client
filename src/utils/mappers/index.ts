import { Banner, Game, Highlight } from 'graphql/types/schema'
import formatPrice from 'utils/formatPrice'

export const bannersMapper = (banners: Banner[]) => {
  return banners.map((banner) => ({
    title: banner.title,
    img: `http://localhost:1337${banner.image?.url}`,
    buttonLabel: banner.button.label,
    buttonLink: banner.button.link,
    subtitle: banner.subtitle,
    ...(banner.ribbon && {
      ribbon: banner.ribbon?.text,
      ribbonColor: banner.ribbon!.color,
      ribbonSize: banner.ribbon!.size
    })
  }))
}

export const gamesMapper = (games: Game[]) => {
  return games.length
    ? games.map((game) => ({
        id: game.id,
        slug: game.slug,
        title: game.name,
        developer: game.developers[0]?.name,
        image: `http://localhost:1337${game.cover?.url}`,
        price: game.price
      }))
    : []
}

export const highlightMapper = (highlight: Highlight) => {
  return (
    highlight && {
      title: highlight.title,
      subtitle: highlight.subtitle,
      alignment: highlight.alignment,
      buttonLink: highlight.buttonLink,
      buttonLabel: highlight.buttonLabel,
      floatImage: `http://localhost:1337${highlight.floatImage?.url}`,
      backgroundImage: `http://localhost:1337${highlight.background?.url}`
    }
  )
}

export const cartItemsMapper = (games: Game[]) => {
  return games.map((game) => ({
    id: game.id,
    title: game.name,
    price: formatPrice(game.price),
    img: `http://localhost:1337${game.cover?.url}`
  }))
}
