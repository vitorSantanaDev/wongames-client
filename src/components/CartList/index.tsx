import Link from 'next/link'

import Button from 'components/Button'
import GameItem from 'components/GameItem'

import { CartListProps } from './types'

import * as S from './styles'

const CartList = ({ items, total, hasButton = false }: CartListProps) => (
  <S.Wrapper>
    {items.map((item) => (
      <GameItem key={item.title} {...item} />
    ))}
    <S.CartListFooter>
      {!hasButton && <span>Total: </span>}
      <S.Total>{total}</S.Total>
      {!!hasButton && (
        <Link href="/cart" passHref>
          <Button as="a">Buy it now</Button>
        </Link>
      )}
    </S.CartListFooter>
  </S.Wrapper>
)

export default CartList
