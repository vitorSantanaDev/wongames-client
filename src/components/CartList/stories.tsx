import { Story, Meta } from '@storybook/react'
import CartList from '.'
import gameItemsMock from './mock'

export default {
  title: 'CartList',
  component: CartList,
  parameters: {
    backgrounds: {
      default: 'won-dark'
    }
  }
} as Meta

export const Default: Story = (args) => (
  <div style={{ maxWidth: '80rem', margin: '0 auto' }}>
    <CartList {...args} />
  </div>
)

Default.args = {
  total: '$440.00',
  cartContextValue: { items: gameItemsMock }
}

export const WithButton: Story = (args) => (
  <div style={{ maxWidth: '80rem', margin: '0 auto' }}>
    <CartList {...args} />
  </div>
)

WithButton.args = {
  hasButton: true,
  total: '$440.00',
  cartContextValue: { items: gameItemsMock }
}

export const Empty: Story = () => (
  <div style={{ maxWidth: '80rem', margin: '0 auto' }}>
    <CartList />
  </div>
)
