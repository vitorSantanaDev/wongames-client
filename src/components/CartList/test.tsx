import { render, screen } from 'utils/test-utils'
import gameItemsMock from './mock'

import CartList from '.'

describe('<CartList />', () => {
  it('should render the heading', () => {
    const totalPriceValue = gameItemsMock.reduce(
      (acc, item) =>
        (acc += Number(item.price.replace('R$ ', '').replace(',', '.'))),
      0
    )
    const { container } = render(
      <CartList items={[...gameItemsMock]} total={`R$ ${totalPriceValue}`} />
    )

    const headings = screen.getAllByRole('heading')
    expect(headings).toHaveLength(2)

    const totalPrice = screen.getByText(`R$ ${totalPriceValue}`)
    expect(totalPrice).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the button', () => {
    const totalPriceValue = gameItemsMock.reduce(
      (acc, item) =>
        (acc += Number(item.price.replace('R$ ', '').replace(',', '.'))),
      0
    )
    render(
      <CartList
        items={[...gameItemsMock]}
        total={`R$ ${totalPriceValue}`}
        hasButton
      />
    )

    const button = screen.getByText(/buy it now/i)
    expect(button).toBeInTheDocument()
  })

  it('should render Empty if there are no items', () => {
    render(<CartList />)
    const emptyDescription = screen.getByText(/your cart is empty/i)
    expect(emptyDescription).toBeInTheDocument()

    const totalLabel = screen.queryByText(/total/i)
    expect(totalLabel).not.toBeInTheDocument()
  })
})
