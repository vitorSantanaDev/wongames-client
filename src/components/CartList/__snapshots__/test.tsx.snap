// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartList /> should render the heading 1`] = `
.c1 {
  padding: 2.4rem;
  border-bottom: 0.1rem solid #EAEAEA;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 1.2rem;
  width: 9.6rem;
  height: 5.6rem;
}

.c3 img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c5 {
  color: #030517;
  font-weight: 600;
  font-size: 1.4rem;
  line-height: 1.4rem;
  margin-bottom: 0.8rem;
}

.c6 {
  color: #FAFAFA;
  font-weight: 600;
  padding: 0.2rem 0.8rem;
  font-size: 1.4rem;
  border-radius: 0.4rem;
  background-color: #3CD3C1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-self: start;
  -ms-flex-item-align: start;
  align-self: start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #FAFAFA;
}

.c7 {
  padding: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  color: #030517;
  font-weight: 600;
  font-size: 1.4rem;
  background-color: #eee;
}

.c8 {
  color: #F231A5;
}

@media (min-width:768px) {
  .c1 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media (min-width:768px) {
  .c3 {
    width: 15rem;
    height: 7rem;
    margin-right: 1.6rem;
  }
}

@media (min-width:768px) {
  .c5 {
    font-size: 2rem;
    line-height: 2rem;
  }
}

@media (min-width:768px) {
  .c7 {
    padding: 2.4rem;
    font-size: 1.6rem;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <img
          alt="Hoziron II"
          src="https://wallpaperaccess.com/full/3149843.jpg"
        />
      </div>
      <div
        class="c4"
      >
        <h3
          class="c5"
        >
          Hoziron II
        </h3>
        <p
          class="c6"
        >
          R$ 350,00
        </p>
      </div>
    </div>
  </div>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <img
          alt="Halo 5"
          src="https://wallpapers-hub.art/wallpaper-images/263763.jpg"
        />
      </div>
      <div
        class="c4"
      >
        <h3
          class="c5"
        >
          Halo 5
        </h3>
        <p
          class="c6"
        >
          R$ 255,00
        </p>
      </div>
    </div>
  </div>
  <div
    class="c7"
  >
    <span>
      Total: 
    </span>
    <span
      class="c8"
    >
      R$ 605
    </span>
  </div>
</div>
`
